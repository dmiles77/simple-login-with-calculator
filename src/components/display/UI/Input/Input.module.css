@import './../../../../variables.css';

.InputStyles {
    width: 98%;
    box-sizing: border-box;
    font-family: inherit;
}

.InputStyles input {
    background-color: rgba(241, 238, 238, 0);
    color: var(--color-main-dark);
    font-size: 1.2rem;
    width: 100%;
    height: 100%;
    padding: 0.3rem 1rem;
    margin: 0;
    display: inline-block;
    transition: all 0.5s;
    outline: none;
}

/* Medium screens */
@media only screen and (max-width: 770px) {
    .InputStyles input::placeholder {
        font-size: var(--text-size-medium-small);
    }
}