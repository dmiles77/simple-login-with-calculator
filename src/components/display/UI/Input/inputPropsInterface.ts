// The Input component prop type TS interface
export interface IInputProps {
	elementConfig: {
		type: string;
		placeholder: string;
		label: string;
	};
	value: string | number;
	isFocused: boolean;
	handleChange?: (event: object) => void;
}
