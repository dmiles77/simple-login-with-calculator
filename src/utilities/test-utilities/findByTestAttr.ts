import { ShallowWrapper } from 'enzyme';

// Used to find a specific DOM element that is specified with the 'data-test' attribute to better traverse nodes while testing:
const findByTestAttr = (component: ShallowWrapper, attr: string) => {
	const wrapper = component.find(`[data-test='${attr}']`);
	return wrapper;
};

export default findByTestAttr;
