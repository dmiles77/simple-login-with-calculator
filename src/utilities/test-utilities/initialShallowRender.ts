import { ShallowWrapper } from 'enzyme';

// Used to quickly test basic shallow wrapper components for render success:
const initialShallowRender = (
	comp: ShallowWrapper,
	className?: string
): void => {
	expect(comp.length).toBe(1);
	expect(comp.length).not.toBe(2);
	if (className) expect(comp).toHaveClassName(`${className}`);
};

export default initialShallowRender;
