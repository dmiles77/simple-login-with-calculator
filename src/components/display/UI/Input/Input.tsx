import React, { useRef, useEffect } from 'react';
import styles from './Input.module.css';
import { IInputProps } from './inputPropsInterface';

const Input: React.FC<IInputProps> = ({
	elementConfig,
	value,
	isFocused,
	handleChange,
}) => {
	const { InputStyles } = styles;

	// Focus the first input field upon component mount
	const focusRef: React.RefObject<any> = useRef();

	useEffect(() => {
		if (isFocused) focusRef.current.focus();
	}, [isFocused]);
	return (
		<div className={InputStyles}>
			<label>{elementConfig.label}</label>
			<input
				ref={focusRef}
				{...elementConfig}
				value={value}
				onChange={handleChange}
				data-test="input-test"
			/>
		</div>
	);
};

export default Input;
