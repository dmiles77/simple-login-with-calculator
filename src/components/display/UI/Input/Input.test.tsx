import React from 'react';
import Input from './Input';
import { IInputProps } from './inputPropsInterface';
import { shallow, ShallowWrapper } from 'enzyme';
import findByTestAttr from '../../../../utilities/test-utilities/findByTestAttr';
import initialShallowRender from '../../../../utilities/test-utilities/initialShallowRender';

const onChange: () => void = jest.fn();

const setComponentProps = (
	configType: string = 'text',
	value: string = 'abc',
	hasInput: boolean = true
) => {
	const props: IInputProps = {
		elementConfig: {
			type: `${configType}`,
			placeholder: 'abc',
			label: 'abc',
		},
		value: `${value}`,
		isFocused: true,
		handleChange: () => {},
	};
	return props;
};

const setComponent = (props: IInputProps, changeFn = () => {}) => {
	const component: ShallowWrapper = shallow(
		<Input {...props} handleChange={changeFn} />
	);
	return component;
};

describe('Input component', () => {
	describe('Component (not testing the input fields)', () => {
		it('should render an Input component withot errors', () => {
			let props = setComponentProps('text', 'abc', true);
			let component = setComponent(props);
			initialShallowRender(component, '.InputStyles');
		});
	});

	describe('Type: input="text" test', () => {
		let props = setComponentProps('input');
		let component = setComponent(props);
		let input = findByTestAttr(component, 'input-test');

		it('should render the component without errors', () => {
			expect(component).toMatchSnapshot();
			expect(input.length).toBe(1);
		});

		it('should call the onChange event callback function on user input', () => {
			props = setComponentProps('input', 'abc', true);
			component = setComponent(props, onChange);
			input = findByTestAttr(component, 'input-test');

			input.simulate('change', {
				target: { value: 'abc' },
			});

			expect(onChange).toHaveBeenCalled();
		});
	});
});
