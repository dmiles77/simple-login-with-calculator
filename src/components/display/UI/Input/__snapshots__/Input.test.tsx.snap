// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Component (not testing the input fields) should render an Input component without errors 1`] = `
<div
  className="InputStyles"
>
  <label>
    abc
  </label>
  <input
    data-test="input-test"
    label="abc"
    onChange={[Function]}
    placeholder="abc"
    type="text"
    value="abc"
  />
</div>
`;

exports[`Input component Type: input="text" test should render the component without errors 1`] = `
<div
  className="InputStyles"
>
  <label>
    abc
  </label>
  <input
    data-test="input-test"
    label="abc"
    onChange={[Function]}
    placeholder="abc"
    type="input"
    value="abc"
  />
</div>
`;
